### PR TITLE
Modernize

### PR DESCRIPTION
**Pull request by winstliu**
Sorta similar to the recent Keytar PR.  I got tired of having to deal with a coffeelint version so old that linter-coffeelint wouldn't even recognize it.

* Replace Coffeescript with JavaScript
* Replace Coffeelint with Standard
* Replace Jasmine with Mocha
* Remove Grunt in favor of npm scripts
* Remove Grim and deprecated methods/events
* Replace underscore-plus in favor of "vanilla" JavaScript.  This makes some portions _slightly_ more verbose but I think it's worth it.
* Move fs-plus to `devDependencies` as it's only being used in one spec
* Fix some docs
* Add benchmarking as part of the testing process
* Move from sudo-enabled Precise to Trusty on Travis

As this is my first experience using decaffeinate, it's very probable that something in this PR is incorrect or is better done a different way.  I verified that all tests passed before converting the tests themselves to Chai assert-style.  Benchmarking before and after didn't reveal any noticeable differences for a non-minified jQuery file.

Some future improvements could be:
* Remove the pegjs dependency and convert `scope-selector-matchers.pegjs` to plain JavaScript
* Improve joanna such that the glob dependency can be removed

### CODE REVIEW

**Comment by Ingramz** [comment]
I wouldn't say that removing PEG.js in favor of a handwritten monstrosity is an improvement, unless PEG.js generated parser is unreasonably slower than a handwritten alternative. I'm currently investigating why first-mate scope selector behaves differently from Textmate's implementation and I have to say that it took me a fraction of time to grok the first-mate parser although Textmate's is not that long either. I believe that at the moment maintainability is still more important than a potentially minuscule performance gain.

However rest of the PR is good, haven't had the chance to run it yet but diffs LGTM.

**Comment by 50Wliu** [comment]
Cool.  Obviously if it turns out that a JS implementation will be unreadable, then I'm fine with keeping it.  I was just hoping that we could get rid of another dependency along the way.

Also I need to push another commit to fix Atom compatibility.

**Comment by Alhadis** [comment]
[Who the hell says handwritten parsers have to be unreadable?](https://github.com/Alhadis/language-roff/blob/c939bbe/lib/tokeniser.js#L128)

File: benchmark/benchmark.js [hunk 1]
```diff
@@ -19,7 +19,7 @@ function tokenize (grammar, content, lineCount) {
     tokenCount += tokenizedLine.length
   }
   const tokensPerMillisecond = Math.round(tokenCount / duration)
-  console.log(`Generated #{tokenCount} tokens for ${lineCount} lines in ${duration}ms (${tokensPerMillisecond} tokens/ms)`)
+  console.log(`Generated ${tokenCount} tokens for ${lineCount} lines in ${duration}ms (${tokensPerMillisecond} tokens/ms)`)
```
Comment (Alhadis) [inline benchmark/benchmark.js]: You might find [this command](https://github.com/Alhadis/Atom-PhoenixTheme/blob/master/src/commands/editor.js#L46-L65) to be helpful. =) It'll automatically prepend a `$` when you insert curly brackets in a JS string literal.

Comment (50Wliu) [inline benchmark/benchmark.js]: What happened here was I copied it verbatim from benchmark.coffee and forgot to change this one.

**Comment by kaled-3000** [comment]
m276051@gmail.com 62.4.3.38

**Comment by kaled-3000** [comment]
m276051@gmail.com 62.4.3.38

**Comment by kaled-3000** [comment]
m276051@gmail.com 62.4.3.38

**Comment by ghost** [comment]
I assume any new PEGjs implementation won't be instantiating `RegExp`? I found that the performance of a regex solution was 15% faster if the majority of the syntax was removed: SNDST00M/vscode-textmate-languageservice#18

**Comment by ghost** [comment]
A perf boost you may want to consider is caching the results of ScopeSelector methods for each `scopes,join(' ')`.

**Comment by winstliu** [comment]
I'm going to close this - I'm way too out of touch with the current state of things to make merging this feasible.
That said, I haven't forgotten about your ask to help out with a TypeScript port (I think it was you who asked?), but I'm not sure when I'll have the time to contribute.

**Comment by ghost** [comment]
I spent multiple hours yesterday just running the test suite and inspecting the PegJS outputs to get that working, and it did

https://github.com/SNDST00M/vscode-textmate-languageservice/commit/4ddf18574ef391f569ae39f29a3915820e23efcc

**Pull request closed without merging**